lib/matplotlib/text.py

```diff
@@ -279,6 +279,41 @@ def _get_multialignment(self):
         else:
             return self._horizontalalignment
 
+    def _char_index_at(self, x):
+        """
+        Calculates the index nearer to x, we accumulate
+        the width of each character contained in the text
+        and find the index at which the distance between the
+        click position and the accumulated sum is minimal.
+        This works only on single line texts.
+        """
+
+        if (not hasattr(self, "_saved_fontproperties") or
+                str(self._fontproperties) != self._saved_fontproperties):
+
+            self._saved_fontproperties = str(self._fontproperties)
+            self._chars_size = dict()  # to cache results
+
+        text = self._text
+
+        if len(text) == 0:
+            return 0
```

Comment:
Style: I would move this to the top of the function so that you have an early return pattern:
```
if not self._text:
    return 0
```
and the standard non-trivial case, which basically needs all the code logic does not need to be indented into an else clause.

Reply:
done! I also added tests